draft-ietf-mls-protocol.md

```diff
@@ -2500,6 +2555,66 @@ that a malicious participant could cause a denial of service by sending a
 handshake message with invalid values for public keys in the ratchet
 tree. -->
 
+### External Commits
+
+External Commits are a mechanism for new members (external parties that want to
+become members of the group) to add themselves to a group, without requiring
+that an existing member has to come online to issue a Commit that references an
+Add Proposal.
+
+Whether existing members of the group will accept or reject an External Commit
+follows the same rules that are applied to other handshake messages.
+
+New members can create and issue an External Commit if they have access to the
+following information for the group's current epoch:
+
+* group ID
+* epoch ID
+* ciphersuite
+* public tree hash
+* confirmed transcript hash
+* group extensions
+* external public key
+
+This information is aggregated in a `GroupKeyPackage` object as follows:
+
+```
+struct {
+    CipherSuite cipher_suite;
+    opaque group_id<0..255>;
+    uint64 epoch;
+    opaque tree_hash<0..255>;
+    opaque confirmed_transcript_hash<0..255>;
+    Extension extensions<0..2^32-1>;
+    HPKEPublicKey external_pub;
+} GroupKeyPackage;
+```
+
+Note that the `tree_hash` field is used the same way as in the Welcome message.
+The full tree can be included via the `ratchet_tree` extension
+{{ratchet-tree-extension}}.
+
+The information above are not deemed public data in general, but applications
+can choose to make them available to new members in order to allow External
+Commits.
+
+External Commits work like regular Commits, with a few differences:
+
+* External Commits MUST reference an Add Proposal that adds the issuing new
+  member to the group
+* External Commits MUST contain a `path` field (and is therefore a "full"
+  Commit)
+* External Commits MUST be signed by the new member.  In particular, the
+  signature on the enclosing MLSPlaintext MUST verify using the public key for
+  the credential in the `leaf_key_package` of the `path` field.
+* An external commit MUST reference exactly one ExternalInit proposal. When
+  processing a Commit, both existing and new members MUST use the external
+  init secret as described in {{external-initialization}}.
```

Comment:
```suggestion
* An external commit MUST reference exactly one ExternalInit proposal, and the
  sender of the proposal MUST be the same as the sender of the Commit. When
  processing a Commit, both existing and new members MUST use the external init
  secret as described in {{external-initialization}}.
```

Reply:
I think we said that `ExternalInit` can only be an inline proposal, but we didn't specify it. So I'd rather specify that requirement in the ExternalInit section.

Reply:
(I updated your suggestion)

Reply:
I was about to argue that I don't think it technically needs to be an inline proposal, but it does -- there's no way to send ExternalInit if you're not in the group.  (You can use `new-member` for Add, but not ExternalInit.)  So your tweak works for me.